compiler/rustc_symbol_mangling/src/v0.rs

```diff
@@ -486,26 +486,26 @@ impl Printer<'tcx> for SymbolMangler<'tcx> {
         predicates: &'tcx ty::List<ty::Binder<'tcx, ty::ExistentialPredicate<'tcx>>>,
     ) -> Result<Self::DynExistential, Self::Error> {
         for predicate in predicates {
-            self = self.in_binder(&predicate, |mut cx, predicate| {
-                match predicate {
-                    ty::ExistentialPredicate::Trait(trait_ref) => {
+            match predicate.as_ref().skip_binder() {
+                ty::ExistentialPredicate::Trait(trait_ref) => {
+                    self = self.in_binder(&predicate, |mut cx, _predicate| {
                         // Use a type that can't appear in defaults of type parameters.
                         let dummy_self = cx.tcx.mk_ty_infer(ty::FreshTy(0));
                         let trait_ref = trait_ref.with_self_ty(cx.tcx, dummy_self);
                         cx = cx.print_def_path(trait_ref.def_id, trait_ref.substs)?;
-                    }
-                    ty::ExistentialPredicate::Projection(projection) => {
-                        let name = cx.tcx.associated_item(projection.item_def_id).ident;
-                        cx.push("p");
-                        cx.push_ident(&name.as_str());
-                        cx = projection.ty.print(cx)?;
-                    }
-                    ty::ExistentialPredicate::AutoTrait(def_id) => {
-                        cx = cx.print_def_path(*def_id, &[])?;
-                    }
+                        Ok(cx)
+                    })?;
                 }
-                Ok(cx)
-            })?;
+                ty::ExistentialPredicate::Projection(projection) => {
```

Comment:
Also, based on @eddyb's [comment](https://github.com/rust-lang/rust/issues/83611#issuecomment-811580990), we should just do a simple validation here.

When we encounter an `ExistentialPredicate::Trait`, we should store that in a local variable `last_trait_ref: Binder<'tcx, TraitRef>` then when we see an `ExistentialPredicate::Projection`, we want to check that 1) the binders match and 2) the projection's trait is the same as the last trait

Reply:
I added some assertions, are they what you expected? 65c04123f81